presto-orc/src/test/java/com/facebook/presto/orc/TestSelectiveOrcReader.java

```diff
@@ -1062,6 +1064,61 @@ public void testOutputNotRequired()
         }
     }
 
+    @Test
+    public void testAdaptiveBatchSizes()
+            throws Exception
+    {
+        Type type = VARCHAR;
+        List<Type> types = ImmutableList.of(type);
+
+        TempFile tempFile = new TempFile();
+        List<String> values = new ArrayList<>();
+        int rowCount = 10000;
+        int longStringLength = 5000;
+        Random random = new Random();
+        long start = System.currentTimeMillis();
+        for (int i = 0; i < rowCount; ++i) {
+            if (i < MAX_BATCH_SIZE) {
+                StringBuilder builder = new StringBuilder();
+                for (int j = 0; j < longStringLength; ++j) {
+                    builder.append(random.nextInt(10));
+                }
+                values.add(builder.toString());
+            }
+            else {
+                values.add("");
+            }
+        }
+        System.out.println(System.currentTimeMillis() - start);
+        writeOrcColumnsPresto(tempFile.getFile(), DWRF, NONE, Optional.empty(), types, ImmutableList.of(values), new OrcWriterStats());
+
+        try (OrcSelectiveRecordReader recordReader = createCustomOrcSelectiveRecordReader(tempFile, OrcEncoding.DWRF, OrcPredicate.TRUE, type, MAX_BATCH_SIZE, false)) {
+            assertEquals(recordReader.getFileRowCount(), rowCount);
+            assertEquals(recordReader.getReaderRowCount(), rowCount);
+            assertEquals(recordReader.getFilePosition(), 0);
+            assertEquals(recordReader.getReaderPosition(), 0);
+
+            // Size of the first batch should equal to the initial batch size (set to MAX_BATCH_SIZE)
+            Page page = recordReader.getNextPage();
+            assertNotNull(page);
+            page = page.getLoadedPage();
+            assertEquals(page.getPositionCount(), MAX_BATCH_SIZE);
+
+            // Later batches should be adjusted based on maxCombinedBytesPerRow collected during the first batch read
+            while (true) {
+                page = recordReader.getNextPage();
+                assertNotNull(page);
+                page = page.getLoadedPage();
+                if (recordReader.getReadPositions() < rowCount) {
```

Comment:
I was staring at this code for some time trying to understand how it works. It might be clearer to move the condition into the while:

```
while (recordReader.getReadPositions() < rowCount) {
   ...
}
```

Reply:
That will require the assert line to be changed to:
```
assertTrue(page.getPositionCount() == MAX_BLOCK_SIZE.toBytes() / (longStringLength + Integer.BYTES + Byte.BYTES) || recordReader.getReadPositions() == rowCount);
```

Do you prefer this way?

Reply:
Is this to cover the last batch?

Reply:
I think I understand. It is ok. Let's merge as is. I restarted the tests as there seems to be some flakiness (CC: @tdcmeehan). Ping me when green so I can merge.